### PR TITLE
Display error banner when Reports server returns status code ≠ 2xx

### PR DESCRIPTION
**Pull request by vincent-psarga**
Other Cucumber implementations display the content send by the server upon failure. For example, with Cucumber ruby:

```
$ CUCUMBER_PUBLISH_TOKEN=plop bundle exec cucumber features/docs/getting_started.feature
WARNING: Nokogiri was built against LibXML version 2.9.10, but has dynamically loaded 2.9.4
Using the default profile...
........

2 scenarios (2 passed)
8 steps (8 passed)
0m0.023s
┌──────────────────────────────────────────────────────────────┐
│ Invalid CUCUMBER_PUBLISH_TOKEN: plop                         │
│ Go to https://reports.cucumber.io/profile to get your token. │
└──────────────────────────────────────────────────────────────┘
Traceback (most recent call last):
	1: from /Users/vincent.pretre/dev/open-source/cucumber-ruby-meta/cucumber-ruby/lib/cucumber/formatter/io.rb:40:in `block (3 levels) in new'
/Users/vincent.pretre/dev/open-source/cucumber-ruby-meta/cucumber-ruby/lib/cucumber/formatter/http_io.rb:79:in `close': request to https://messages.cucumber.io/api/reports failed with status 401 (StandardError)
```

This is an attempt to get the same behavior in `cucumber-js`

TODO:
 - [x] Add a example of what's excepted
 - [x] Display the server response
 - [x] Fail when the server does not return 2xx
 - [ ] Solve the `null` line displayed above the banner

### CODE REVIEW

**Comment by vincent-psarga** [comment]
The build is not really stable (I managed to get it green by relaunching - and after it failed again).

Apparently that comes from some tests I have added and the way the stream errors are handled. I'm going to dig a bit more into that (although the solution I have in mind is not that beautiful to be honest :D )

**Comment by aslakhellesoy** [comment]
@vincent-psarga are the non-deterministic tests new tests in this PR? I can help have a look at this.

**Comment by davidjgoss** [comment]
The flaky tests I think started from the original change where publish support was added per #1457

I tried a few things a while ago but hit a bit of a wall tbh.

File: .github/workflows/build.yml
Comment (davidjgoss) [inline .github/workflows/build.yml:19]: Agree with keeping this!

Comment (vincent-psarga) [reply .github/workflows/build.yml:19]: great :) I'll remove the comment then :)

**Comment by vincent-psarga** [comment]
> @vincent-psarga are the non-deterministic tests new tests in this PR? I can help have a look at this.

At least from what I see, it's the test that I've introduced which is flaky. I'm looking at a way to not spawn Cucumber in order to fix this (worked fine with Ruby getting rid of spawning, so I give a try)

**Comment by aurelien-reeves** [comment]
Just tried to remove the delays in the server
The error went back

So it seems the combination of the delays, and the ports, stabilized the tests.

**Comment by aurelien-reeves** [comment]
@davidjgoss do you have any new feedback regarding that PR?
We would like to merge it and release it.

I plan to open a new PR dedicated to the inconsistent test once that one is merged.

**Comment by davidjgoss** [comment]
@aurelien-reeves I'll review this morning, thanks for the nudge!

File: features/publish.feature
Comment (davidjgoss) [inline features/publish.feature:108]: As a general point, not sure how users would feel about their CI pipeline failing because e.g. cucumber reports is down or they got their token wrong (but all tests passed). We could wait for feedback and look at making it configurable via formatter options if needed.

**Review by davidjgoss — APPROVED** [review]
LGTM

File: features/publish.feature
Comment (vincent-psarga) [reply features/publish.feature:108]: Well, that's already the case if you publish with a wrong token, `cucumber-js` exits with code 1. What this PR does is just providing extra information.
And I think it should fail if the publication did not succeed, as something went wrong. It a bit like generating a JUnit report but the output directory can't be written, you would expect the run to exit with an error no ?

Comment (aurelien-reeves) [reply features/publish.feature:108]: I agree with @vincent-psarga 
People who expect reports to be published from a CI for example should be noticed that something went wrong.

I run a few tests with cucumber-ruby. I was expecting a non-zero exit code, it was not the case. I think I'll open an issue in cucumber-ruby about that.

I will try the same with cucumber-jvm

Comment (davidjgoss) [reply features/publish.feature:108]: Makes sense to me. I think it’s fine to kick the option down the road and see if there’s demand.

**Pull request merged**
